mne_bids/path.py

```diff
@@ -120,22 +132,88 @@ def __str__(self):
                     key = 'rec'
                 basename.append('%s-%s' % (key, val))
 
-        if self.suffix is not None:
-            basename.append(self.suffix)
+        if self.kind is not None:
+            if self.extension is not None:
+                basename.append(f'{self.kind}{self.extension}')
+            else:
+                basename.append(self.kind)
 
         basename = '_'.join(basename)
-        if self.prefix is not None:
-            basename = op.join(self.prefix, basename)
-
         return basename
 
+    @property
+    def fpath(self):
+        """Get the full filepath for this BIDS file.
+
+        Getting the file path consists of the following behavior:
+
+        1. ``bids_root`` is not passed into the object: A warning is shown
+        to the user and it return the ``basename``.
+
+        2. ``bids_root`` is passed into the object:
+
+            i. If ``kind`` and ``extension`` are given:
+            The function will then return the full filepath
+            using ``bids_root`` and ``basename``
+
+            ii. Else:
+            The function tries to infer the full filepath
+            for the given ``bids_root`` and ``basename``
+            and will return the full filepath. It will try to
+            infer the filename extension by searching
+            for the file on disk. If the file cannot be found, an error
+            will be raised. To disable this automatic inference attempt,
+            pass a string (like ``'.fif'`` or ``'.vhdr'``).
+
+            If it cannot find a matching dataset, it
+            will result in an error.
+
+        If ``None``,
+
+        Returns
+        -------
+        bids_fpath : str
+            Either the ``basename``, or full filepath to the dataset.
+        """
+        if self.bids_root is None:
+            msg = ('No bids root was passed in and '
+                   'full BIDS filepath cannot be '
+                   'automatically inferred. '
+                   'Returning the bids basename.')
+            warn(msg)
+            return self.basename
+
+        # TODO: how do we get rid of this? Hack to make sure no error out
+        # during write/read.
```

Comment:
what's the hack? I don't follow

Reply:
This is unfortunately something I am still trying to figure out. The problem with `fpath` property is that it allows you to infer what a file path is based on if it exists or not.

However, what occurs in `write_raw_bids` is that some of these BIDSPath's don't exist yet, so calling `fpath` would error out. 

We could:
1. refactor `write_raw_bids` to get the explicit filepaths before passing it into the corresponding write functions (e.g. for events, channels, etc.)
2. rethink how to allow `fpath` that doesn't exist yet 

Any suggestions on how to proceed?